tictactoe/src/main/java/models/TicTacToeService.java

```diff
@@ -0,0 +1,144 @@
+      int inputColumnNum = gameboard.getUserRowColumnInputs().get(k).getInputColumnNum();
+      if(!isValidMove(grid,inputRowNum,inputColumnNum)) {
+        System.out.println("Invalid Move");
+        playerQueue.add(0,currentPlayer);
+      }
+      else {
+        grid[inputRowNum-1][inputColumnNum-1] = currentPlayer.getChosenCharacter()+" ";
+        displayGrid(grid);
+        if(isGameComplete(grid)) {
+          System.out.println("Game Over");
+          break;
+        }
+        if(rowCheck(grid,inputRowNum,inputColumnNum) || columnCheck(grid,inputRowNum,inputColumnNum)
+            || leftDiagonalCheck(grid,inputRowNum,inputColumnNum) || rightDiagonalCheck(grid,inputRowNum,inputColumnNum)) {
+          System.out.println(currentPlayer.getName()+" won the game");
+          break;
+        }
+        playerQueue.add(currentPlayer);
+      }
+    }
+  }
+
+  private void displayGrid(String[][] grid) {
+    for (String[] strings : grid) {
+      for (String string : strings) {
+        System.out.print(string);
+      }
+      System.out.println();
+    }
+  }
+
+  private boolean isValidMove(String [][]grid,int inputRowNum,int inputColumnNum) {
+    return grid[inputRowNum - 1][inputColumnNum - 1].equals("- ") && (inputRowNum >= 1
+        && inputRowNum <= grid.length)
+        && (inputColumnNum >= 1 && inputColumnNum <= grid.length);
+  }
+
+  private boolean columnCheck(String [][]grid,int inputRowNum,int inputColumnNum) {
+    for(int i=0;i<grid.length;i++) {
+      if(!grid[inputRowNum - 1][i].equals(grid[inputRowNum - 1][inputColumnNum - 1])) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  private boolean rowCheck(String [][]grid,int inputRowNum,int inputColumnNum) {
+    for (String[] strings : grid) {
+      if (!strings[inputColumnNum - 1].equals(grid[inputRowNum - 1][inputColumnNum - 1])) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  private boolean leftDiagonalCheck(String [][]grid,int inputRowNum,int inputColumnNum) {
+    int characterCounter=0;
+    for(int i=0;i<grid.length;i++) {
+      for(int j=0;j<grid.length;j++) {
+        if(i==j && grid[i][j].equals(grid[inputRowNum - 1][inputColumnNum - 1])) {
+          characterCounter++;
+        }
+      }
+    }
+    return characterCounter == grid.length;
+  }
+
+  private boolean rightDiagonalCheck(String [][]grid,int inputRowNum,int inputColumnNum) {
+    int characterCounter=0;
+    for(int i=grid.length-1;i>=0;i--) {
+      for(int j=grid.length-1;j>=0;j--) {
+        if(i==j && grid[i][j].equals(grid[inputRowNum - 1][inputColumnNum - 1])) {
+          characterCounter++;
+        }
+      }
+    }
+    return characterCounter == grid.length;
+  }
+
+  private boolean isGameComplete(String [][]grid) {
+    int xCounter = 0;
+    int oCounter = 0;
+    int noOfRowsGrid = grid.length;
+    for (String[] strings : grid) {
+      for (int j = 0; j < grid.length; j++) {
+        if (strings[j].equals("X ")) {
```

Comment:
why is there an extra space?

These should be constants.

Reply:
Here just to add extra space between the characters, while displaying the entire grid

Reply:
Mixing the display logic with the core logic? Bad practice

Reply:
will take care of it, next time